<a href="https://colab.research.google.com/github/heysatya/DeepLearningKeras/blob/main/Graded_Assignment_Project_Regression_Model_in_Keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 align=center><font size = 5>Graded Assignment Project - Regression Model in Keras</font></h1>


## Introduction


In this course project, I will build a regression model using the deep learning Keras library, and then experiment with increasing the number of training epochs and changing number of hidden layers to find out how changing these parameters impacts the performance of the model.


<h4>Objective for this Assignment<h4>    
<h5> A. Build a baseline model </h5>
<h5> B. Normalize the data </h5>
<h5> C. Increase the number of epochs </h5>
<h5> D. Increase the number of hidden layers </h5>   


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
<a href="#item31">A. Build a baseline model</a>  
<a href="#item32">B. Normalize the data</a>  
<a href="#item33">C. Increase the number of epochs </a>  
<a href="#item34">D. Increase the number of hidden layers </a>  

</font>
</div>


<a id="item31"></a>


## A. Build a baseline model

In this first step, we will use the Keras library to build a neural network with the following:
- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error  as the loss function.


#### Import the Pandas and the Numpy libraries.


In [1]:
import pandas as pd
import numpy as np

The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. 
    
The data has been downloaded from: https://cocl.us/concrete_data.
    
The predictors in the data of concrete strength include include:
1. Cement
2. Blast Furnace Slag
3. Fly Ash
4. Water
5. Superplasticizer
6. Coarse Aggregate
7. Fine Aggregate


#### Read the data into a Pandas dataframe.


In [2]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

As can be seen from the data, the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. This concrete mix is 28 days old and has a compressive strength of 79.99 MPa. 


#### Split data into predictors and target


The target variable in this case is the concrete sample strength. Therefore, all the columns will be considered as predictors.


In [3]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>


#### Sanity check of the predictors and the target dataframes.


In [4]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [5]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

#### Save the number of predictors to _n_cols_ since we will need this number later on.


In [6]:
n_cols = predictors.shape[1] # number of predictors

#### Import the Keras library


In [7]:
import keras

#### Import the rest of the packages from the Keras library that we will need to build the regressoin model.


In [8]:
from keras.models import Sequential
from keras.layers import Dense

### Build a Neural Network


Define a function that defines the regression model that can be conveniently called it to create our model. Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error  as the loss function.


In [9]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

<a id="item4"></a>


### Train and Test the Neural Network

#### Create the model.


In [10]:
# build the model
model = regression_model()

#### 1. Randomly split the data into a training and test sets by holding 30% of the data for testing, using the train_test_split helper function from scikit-learn


In [11]:
from sklearn.model_selection import train_test_split

In [12]:
Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)

#### 2. Train the model on the training data using 50 epochs.

In [13]:
model.fit(Xtrain, ytrain, epochs=50, verbose=1)

Epoch 1/50
23/23 [==============================] - 1s 1ms/step - loss: 220821.1497
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 109642.7008
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 51824.8340
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 24682.2611
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 11554.7799
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 6118.6179
Epoch 7/50
23/23 [==============================] - 0s 6ms/step - loss: 4165.2732
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 3442.0602
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 3093.6308
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 2967.4025
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 2765.7836
Epoch 12/50
23/23 [==============================] - 0s 1ms/step - loss: 2606.6353
Epoch 

#### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength, using the mean_squared_error function from scikit-learn.

In [14]:
from sklearn.metrics import mean_squared_error

In [15]:
MSE = mean_squared_error (ytest, model.predict(Xtest))
print ("MSE between predicted and actual concrete strength: ", MSE)

MSE between predicted and actual concrete strength:  189.92326194746306


#### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors

In [16]:
# build the model
model = regression_model()

In [17]:
MSE = np.empty(50)
for i in range(50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=50, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part A: List of Mean Squared Errors: ", MSE)

Part A: List of Mean Squared Errors:  [204.62386023 117.57650823  92.1441848  127.20617894 109.16836379
  85.80755447  72.87554873 101.7964256   49.83241802  56.77296298
  51.9245604   52.86710782  54.62628687  60.63462231  46.86303951
  49.60715631  50.58064404  58.53309742  50.40535078  51.64528307
  48.41575004  54.41002769  45.05263344  48.26211793  46.71014838
  53.14018732  55.95826397  49.7014904   42.53386846  51.365661
  56.5142754   54.02376332  49.56887522  46.48954268  47.77937741
  41.61425377  47.06105842  45.76400682  47.73003207  56.42465018
  47.80587435  55.12436319  48.928889    47.98280485  49.05640859
  47.24180648  45.45593326  46.75819352  46.67925785  51.70327309]


#### 5. Report the mean and the standard deviation of the mean squared errors.

In [18]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part A: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part A: Standard Deviation of the Mean Squared Errors: ", MSE_std)

Part A: Mean of the Mean Squared Errors:  60.41495884758537
Part A: Standard Deviation of the Mean Squared Errors:  28.117135586677158


## B. Normalize the data

Repeating Part A using a normalized version of the data

Q) How does the mean of the mean squared errors compare to that from Step A?


#### Normalize the data by substracting the mean from the individual predictors and dividing by the standard deviation.

In [19]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

In [20]:
n_cols = predictors_norm.shape[1] # number of predictors

#### Build the model for Part B using normalized data
model = regression_model()

#### Repeat steps in Part A to create a list of 50 mean squared errors (using normalized data model)

In [21]:
MSE = np.empty(50)
for i in range(50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=50, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part B: List of Mean Squared Errors: ", MSE)

Part B: List of Mean Squared Errors:  [35.97349495 46.79673077 51.41033353 56.36646451 49.97935578 47.61795859
 43.90224749 43.47058795 45.53689852 44.18646717 49.63780913 47.61260562
 53.63038179 41.47941658 42.7383987  44.05558029 45.02135105 49.9977054
 50.33390525 47.05174755 47.38489214 54.225938   46.08559216 49.27092801
 42.12879194 42.57838281 39.5871525  41.439141   47.46798377 45.68256563
 47.75398365 48.92418583 53.26483494 65.04573596 41.72506877 46.72745943
 43.06792811 39.84358992 46.17479532 53.75281636 49.4984579  48.17028091
 41.96537463 47.086767   40.85014001 43.99223453 52.39888475 47.57791673
 45.92033309 43.8393963 ]


#### Calculating mean and standard deviation for Part B

In [22]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part B: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part B: Standard Deviation of the Mean Squared Errors: ", MSE_std)

Part B: Mean of the Mean Squared Errors:  46.804619854084834
Part B: Standard Deviation of the Mean Squared Errors:  4.941348110954295


## C. Increase the number of epochs

Repeat Part B using 100 epochs this time for training.

Q) How does the mean of the mean squared errors compare to that from Step B?

#### Repeating steps in Part B using 100 epochs for training

#### Build the model for Part C
model = regression_model()

In [23]:
MSE = np.empty(50)
for i in range(50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=100, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part C: List of Mean Squared Errors: ", MSE)

Part C: List of Mean Squared Errors:  [49.01244807 43.33304682 43.07213573 49.61595151 37.89082494 44.43029875
 40.14412828 40.44547089 42.86058228 44.99583399 41.48051974 43.60020119
 42.53268636 45.81725541 43.26584844 38.31213379 42.87897494 40.75871838
 42.90387156 40.52969693 44.55669794 36.96015647 38.37027081 40.41617238
 35.09841415 46.33142113 44.3256339  30.96938962 43.7954154  42.74130468
 44.02987844 33.25976339 34.89516592 42.99191864 39.03944582 40.35050443
 40.99177902 37.17677183 39.79349745 42.91037882 41.50202938 39.01404397
 42.52232649 40.13749437 32.96367523 41.70437986 46.72815174 37.66590097
 37.9589559  39.41735196]


#### Calculating mean and standard deviation for Part C

In [24]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part C: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part C: Standard Deviation of the Mean Squared Errors: ", MSE_std)

Part C: Mean of the Mean Squared Errors:  41.12997836299742
Part C: Standard Deviation of the Mean Squared Errors:  3.8102513582776254


## D. Increase the number of hidden layers

Repeat part B using a neural network with the following:

- Three hidden layers, each of 10 nodes and ReLU activation function.

Q) How does the mean of the mean squared errors compare to that from Step B?

#### Define regression model with 3 hidden layers, each of 10 nodes and ReLU activation functions


In [31]:
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

#### Build the model for Part D


#### Repeating steps in Part B using 50 epochs for training the neural network of Part D

In [33]:
MSE = np.empty(50)
for i in range(50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=50, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part D: List of Mean Squared Errors: ", MSE)

Part D: List of Mean Squared Errors:  [121.17456501  88.77169546  95.94539329  69.77147163  71.22014853
  69.55994493  72.28442872  60.65206323  55.36719962  44.09799091
  45.41963288  48.18680248  52.21848433  55.07457483  46.32345614
  48.08749686  39.14556331  43.62201889  42.054671    40.82747687
  37.20777878  42.36374575  35.9790727   36.93209218  37.66414697
  52.19097611  46.38419056  42.18759172  37.62044214  37.90793356
  37.85515848  44.74135408  44.23925652  49.59562314  42.57456471
  47.64002371  43.19374526  40.61985735  43.92339043  40.95479115
  44.35269371  39.5065515   37.50644475  41.94618834  43.62735563
  40.14080867  35.0205706   42.85129725  50.42620852  47.2075316 ]


#### Calculating mean and standard deviation for Part D

In [34]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part D: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part D: Standard Deviation of the Mean Squared Errors: ", MSE_std)

Part D: Mean of the Mean Squared Errors:  49.48332929559162
Part D: Standard Deviation of the Mean Squared Errors:  16.351410934698176
